units/src/amount.rs

```diff
@@ -1158,7 +1158,22 @@ impl ops::DivAssign<u64> for Amount {
 impl FromStr for Amount {
     type Err = ParseError;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> { Amount::from_str_with_denomination(s) }
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let result = Amount::from_str_with_denomination(s);
+
+        match result {
+            Err(ParseError::MissingDenomination(_)) => {
+                let d = Amount::from_str_in(s, Denomination::Satoshi);
+
+                if d == Ok(Amount::ZERO) {
+                    Ok(Amount::ZERO)
+                } else {
+                   result
+                }
+            },
+            _ => result
```

Comment:
I'm amazed the compiler doesn't complain about this being moved.

Reply:
This is pretty clever of the borrow checker. I bet there is some long PR/issue somewhere with people arguing about how the borrow checker should be able to work out that the error (non-`Copy`) is never used other than in paths that just return it.

Reply:
A &str slice is a read-only copy and doesn't actually own the data like a heap pointer would.  I think these pointers are actually just to read only parts of the binary itself.  Could be wrong or just making stuff up though.

Reply:
A wait this returns Amount not &str

Reply:
It returns `Result` which contains `Error`, which is not `Copy` IIRC.

I think it has something to do with the `_` pattern.

Reply:
Err yes, it returns result.  I had meant to return to this conversation yesterday but got distracted.  Anyway, Result implements Copy: https://doc.rust-lang.org/1.81.0/src/core/result.rs.html#523.  I'm not sure why `_` would matter in this case.